### ISSUE TITLE
Proposal: Future addListenerIfAbsent

### ISSUE THREAD

**Issue by mickare** [netty/netty#5352]
#### Context:
Currently [addListener](http://netty.io/4.1/api/io/netty/util/concurrent/Future.html#addListener(io.netty.util.concurrent.GenericFutureListener)) appends the listener to an array. There can be duplicates.

In my case, I want to add multiple times (e.g. x 1000) the same close listener to a channel but I don't want this listener to be called multiple times.
Also I don't want to remove the listener prior to adding it. I could do this but I don't like the wasteful [Java array copy](https://github.com/netty/netty/blob/4.1/common/src/main/java/io/netty/util/concurrent/DefaultFutureListeners.java#L62) calls.

My proposal is to add a new method `addListenerIfAbsent` that adds only a listener if it is not already added.

```Java
@Test
  public void testAddListenerIfAbsent() {

    final Promise<Void> promise = new DefaultPromise<Void>(ImmediateEventExecutor.INSTANCE);
    final AtomicInteger count = new AtomicInteger(0);

    promise.addListener(f -> count.incrementAndGet());

    GenericFutureListener<? extends Future<? super Void>> upcount = f -> count.incrementAndGet();
    promise.addListenerIfAbsent(upcount);
    promise.addListenerIfAbsent(upcount);

    promise.setSuccess(null);

    assertEquals(2, count.get());
  }
```

Greetings,
Michael

**Comment by Scottmitch** [comment]
It sounds like you are concerned with performance. If you rely upon the `Future` implementation this may incur some overhead (synchronization, array iteration, etc..) that could be otherwise avoided if it was tracked at a level above the future where more information is known about the threading model. Would handling it at the application layer be a good work-around (while this feature does not yet exist)?

>  I want to add multiple times (e.g. x 1000) the same close listener to a channel

Can you explain why this occurs so many times.

**Comment by mickare** [comment]
Yes, I handled it simple. I am adding now the listener one time in advance as soon as the connection is active, even if it does not matter. In retroperspective this may be better than my initial design because addListener is not called.

I am writing a simple RPC framework. All pending response futures should be cancelled if the connection closes. The initial implementation was quick and dirty so I stumbled upon this.

```Java
private static final AtomicLong CALL_ID_COUNTER = new AtomicLong(0);

private final LoadingCache<Channel, Cache<Long, SettableFuture<Response>>> pendingCalls = ...;
// pendingCalls has RemovalListener that set the SettableFuture.

private final ChannelFutureListener invalidateListener = f -> {
  pendingCalls.invalidate(f.channel);
};

public ListenableFuture<?> call(Channel channel, String procedure, String[] arguments) {
  Future<?> f = SettableFuture.create();
  pendingCalls.getUnchecked(channel).put(CALL_ID_COUNTER.incrementAndGet(), f);
  channel.closeFuture().addListener(invalidateListener);
  sendCall(channel, procedure, arguments, f);
  return f;
}
```
Please be aware that I am using Google Guava here for Caches and Futures. The netty future is used here at `channel.closeFuture()`.

In this case `invalidateListener` would be added multiple times to the close future. In fact the array holds as many entries as `call` is executed.

##### Use-Case of proposal

Add a fixed listener to a future, instead of multiple instances or duplicates.
```Java
private final GenericFutureListener<?> listener = f -> {
  // expensive computation
};

Future<?> f = ...;
f.addListenerIfAbsent(listener);
f.addListenerIfAbsent(listener);
f.addListenerIfAbsent(listener);
f.addListenerIfAbsent(listener);
```
The listener is executed only one time when the future f completes. So the expensive computation is not repeatedly executed.

With the Java 8 standard API and [Map.putIfAbsent(K key, V value)](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#putIfAbsent-K-V-) in mind I thought a method `addListenerIfAbsent` would solve it.

##### Listeners that could benefit
- [static channel listeners](http://netty.io/4.1/api/index.html?io/netty/channel/ChannelFutureListener.html)

Greetings,
Michael

**Comment by Scottmitch** [comment]
>  In retroperspective this may be better than my initial design because addListener is called only once in total per channel.

I don't have all the details but doing the single add does seem preferable in this case.

**Comment by normanmaurer** [comment]
Honestly I don't think we need this
